lib/jido/agent/server.ex

```diff
@@ -1,32 +1,86 @@
 defmodule Jido.Agent.Server do
   @moduledoc """
-  GenServer implementation for managing agent processes.
+  Agent Server implementation using GenStateMachine for robust state management.
 
-  This server handles the lifecycle and runtime execution of agents, including:
-  - Agent initialization and startup
+  This module provides the core server functionality for Jido agents, handling:
+  - Agent lifecycle management with explicit state transitions  
   - Signal processing and routing
-  - State management and transitions
-  - Child process supervision
-  - Graceful shutdown
+  - Instruction execution
+  - Queue management with backpressure
+  - Automatic state validation and debugging
 
-  The server can be started in different modes (`:auto` or `:manual`) and supports
-  both synchronous (call) and asynchronous (cast) signal handling.
+  ## Agent States
+
+  - `:initializing` - Server is starting up and mounting callbacks
+  - `:idle` - Server is ready to accept and process signals
+  - `:planning` - Server is preparing action execution plans
+  - `:running` - Server is actively executing instructions
+  - `:paused` - Server execution is temporarily suspended
+
+  ## State Transitions
+
+  ```
+  initializing -> idle        (after successful initialization)
+  idle         -> planning    (when planning mode is activated)
+  idle         -> running     (when executing instructions directly)  
+  planning     -> running     (when plan execution begins)
+  planning     -> idle        (when planning is cancelled)
+  running      -> paused      (when execution is paused)
+  running      -> idle        (when execution completes)
+  paused       -> running     (when execution resumes)
+  paused       -> idle        (when execution is cancelled)
+  ```
+
+  ## Operation Modes
+
+  - `:auto` - Signals are processed automatically as they arrive
+  - `:step` - Signals require manual processing (useful for debugging)
+
+  ## Usage
+
+  ```elixir
+  # Start an agent server
+  {:ok, pid} = Jido.Agent.Server.start_link(agent: MyAgent)
+
+  # Send synchronous instruction
+  instruction = Jido.Instruction.new!(%{action: MyAction, params: %{key: "value"}})
+  {:ok, result} = Jido.Agent.Server.call(pid, instruction)
+
+  # Send asynchronous signal  
+  {:ok, signal_id} = Jido.Agent.Server.cast(pid, instruction)
+
+  # Get current state
+  {:ok, state} = Jido.Agent.Server.state(pid)
+  ```
   """
 
-  use ExDbug, enabled: false
-  use GenServer
+  use GenStateMachine, callback_mode: :handle_event_function
+
+  # Define a compile-time debug flag to avoid dialyzer issues
+  @debug_enabled Application.compile_env(:jido, :debug_enabled, false)
+
+  # Use a simple conditional debug macro instead of ExDbug to avoid dialyzer issues
+  if @debug_enabled do
+    require Logger
+
+    defp dbug(message, opts \\ []) do
+      Logger.debug("[#{__MODULE__}] #{message}", opts)
+    end
+  else
+    defp dbug(_message, _opts \\ []), do: :ok
+  end
```

Comment:
I don't love this. If I use, `use ExDbug, enabled: false` it throws a bunch of warnings since we're using some function args in `dbug/1` (when we disable dbug, it's like the function arg is unused). If we enabled `ExDbug` then that causes a ton of dialyzer errors.

Reply:
Ok. We can stop using on this PR and have a separate one to remove globally. Still good to keep observability in mind. Perhaps improving telemetry